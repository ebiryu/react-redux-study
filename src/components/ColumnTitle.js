import React from 'react';
import PropTypes from 'prop-types';

class ColumnTitle extends React.Component {
  constructor(props) {
    super(props);
    this.focused = React.createRef();
  }
  componentDidUpdate() {
    if (this.focused.current) {
      this.focused.current.focus();
    }
  }
  render() {
    const {isTitleEditable, columnNumber, onClickColumnTitle, editColumnTitle, columnTitle} = this.props;
    return isTitleEditable 
      ? 
      <form onSubmit={() => onClickColumnTitle(columnNumber)}>
        <input className="column__title"
          onBlur={() => onClickColumnTitle(columnNumber)}
          onChange={event => editColumnTitle(columnNumber, event.target.value)}
          value={columnTitle}
          ref={this.focused}
        />
      </form>
      : 
      <span className="column__title" onClick={() => onClickColumnTitle(columnNumber)}>
        {columnTitle}
      </span>
      
  }
}

ColumnTitle.propTypes = {
  isTitleEditable: PropTypes.bool.isRequired,
  columnNumber: PropTypes.number.isRequired,
  onClickColumnTitle: PropTypes.func.isRequired,
  editColumnTitle: PropTypes.func.isRequired,
  columnTitle: PropTypes.string.isRequired,
};

export default ColumnTitle;