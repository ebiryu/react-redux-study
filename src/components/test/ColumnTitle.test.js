import React from 'react'
import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import ColumnTitle from '../ColumnTitle'

Enzyme.configure({ adapter: new Adapter })

describe('Column Title ', () => {
  let testFn

  beforeEach(() => {
    testFn = jest.fn()
  })

  it('子コンポーネントが存在すること(not editable)', () => {
    const columnTitle = shallow(<ColumnTitle isTitleEditable={false}/>)
    expect(columnTitle.find('span')).toHaveLength(1)
    expect(columnTitle.find('form')).toHaveLength(0)
  })
  it('子コンポーネントが存在すること(editable)', () => {
    const columnTitle = shallow(<ColumnTitle isTitleEditable={true}/>)
    expect(columnTitle.find('span')).toHaveLength(0)
    expect(columnTitle.find('form')).toHaveLength(1)
  })
  it('初期値 (expected "title")', () => {
    const columnTitle = shallow(<ColumnTitle columnTitle={'title'}/>)
    expect(columnTitle.text()).toEqual('title')
  })
  it('クリック時', () => {
    const columnTitle = shallow(<ColumnTitle isTitleEditable={false} onClickColumnTitle={testFn}/>)
    columnTitle.find('span').simulate('click')
    // クリックしても'span'のまま。reduxには影響しない模様。
    expect(columnTitle.find('span').exists()).toBe(true)
    expect(testFn).toHaveBeenCalled()
  })
  it('submit時', () => {
    const columnTitle = shallow(<ColumnTitle isTitleEditable={true} onClickColumnTitle={testFn}/>)
    columnTitle.find('form').simulate('submit')
    expect(testFn).toHaveBeenCalled()
  })
  it('blur時', () => {
    const columnTitle = shallow(<ColumnTitle isTitleEditable={true} onClickColumnTitle={testFn}/>)
    columnTitle.find('input').simulate('blur')
    expect(testFn).toHaveBeenCalled()
  })
  it('titleを入力時', () => {
    const columnTitle = shallow(<ColumnTitle isTitleEditable={true} editColumnTitle={testFn}/>)
    columnTitle.find('input').simulate('change', {target: {value: '7'}})
    expect(testFn).toHaveBeenCalled()
  })
  it('autofocus', () => {
    const columnTitle = mount(<ColumnTitle isTitleEditable={true}/>)
    const { focused } = columnTitle.instance()
    jest.spyOn(focused.current, 'focus')

    columnTitle.instance().componentDidUpdate()
    expect(focused.current.focus).toHaveBeenCalledTimes(1)
  })

  it('snapshot', () => {
    const columnTitle = shallow(<ColumnTitle isTitleEditable={true} onClickColumnTitle={testFn}/>)
    expect(columnTitle).toMatchSnapshot()
  })
})