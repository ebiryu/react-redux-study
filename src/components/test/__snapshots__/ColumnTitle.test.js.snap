// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column Title  snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ColumnTitle
    isTitleEditable={true}
    onClickColumnTitle={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <input
        className="column__title"
        onBlur={[Function]}
        onChange={[Function]}
        value={undefined}
/>,
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "column__title",
        "onBlur": [Function],
        "onChange": [Function],
        "value": undefined,
      },
      "ref": Object {
        "current": null,
      },
      "rendered": null,
      "type": "input",
    },
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <input
          className="column__title"
          onBlur={[Function]}
          onChange={[Function]}
          value={undefined}
/>,
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "column__title",
          "onBlur": [Function],
          "onChange": [Function],
          "value": undefined,
        },
        "ref": Object {
          "current": null,
        },
        "rendered": null,
        "type": "input",
      },
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
